Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,249 @@
+   The sizes of these additional stacks are defined at creation time, but allocation of physical pages happens on-demand.
+6. as also decribed in the previous points, the kernel uses various optimisations for dealing with processes on-demand mapping:
+   1. a physical page is allocated and mapped to a process only when the process accesses it, otherwise it might not be present:
+      1. file backed pages are allocated/mapped only when read/written to
+      2. anonymous pages are allocated only when written to
+   2. zero page: when a page is known to be empty, it is not reserved and mapped; instead,
+      the kernel has one specific (dummy) memory page that is mapped tothe process as read-only
+   3. shared libraries:
+      when the same library pages are mapped by multiple processes, the library physical pages
+      are mapped as read-only into each process address space
+   4. copy-on-write:
+      1. when a library has own data, this is initially mapped as read-only and shared;
+         only when written to, then a separate physical page is reserved for each writing process (threads sahre the same)
+      2. same happens for data pages that were treated as zero-page, but then are written to.
+   5. folios (see also kernel section): data structures that try to better abstract compound pages and *might* also be used
+      to represent optimised contiguous pages on ARM64 (instead of mapping 16 entries, it is possible
+      to map a 16-pages chunk of physically contiguous memory, that is also aligned to a 16 pages
+      boundary, reducing TLB use).
+      1. A folio acts as intermediary between vma and lower level memory management
+      2. it might pre-allocate/map more pages than explicitly requested
+7. read-ahead: when asked to fetch data from disk, the kernel might attempt to optimise the operation,
+   reading more pages than requested, under the assumption that more requests might be coming soon
+ 
+
+#### **Safety-Oriented consideration**
+The following considerations are of a more deductive nature.
+
+1. a process that is supposed to support safety requirements should  not have pages swapped out / dropped / missing,
+   because this would introduce:
+   1. uncertainty in the timing required to recover the content, if not immediately available
+   2. additional risk, involving the userspace paging mechanisms in the fulfilling of the safety requirements
+   3. additional dependency on runtime linking, in case the process requires it, and code pages have been
+      discarded - reloading them from disk will not be sufficient
+2. The optimisations made by the kernel in providing physical backing to process memory make it very
+   questionable if it can be assessed when a (part of) a process memory content is actually present in the
+   system physical memory.
+3. by default, it is to be expected that a process will be exposed to various types of interference from the kernel:
```

Comment:
Above I disagree on using the term "by default". Instead I would say "Unless the Kernel is qualified against a non-interence claim, it is to be expected ...."

Reply:
Sure you can disagree, but what evidence is there that the kernel will not interfere?
Qualification is a bit misleading though, because it is not clear on which ground.
You can have a "qualified" kernel and still incur in interference, because it was not proven that it cannot interfere, and the "qualification" happens to be incorrect.

"Qualified according to ISO26262" would be more acceptable, for example.

But here I'm really trying to stick to objective evidence.

A piece of paper will not stop a bug - the MMU (for example) might.

Reply:
```suggestion
3. we cannot rule out the possibility that a process will be exposed to various types of interference from the kernel:
```